Regression for enums under usage under CC

Based on the OpenCB failure in lampelf/steps - [build logs](https://scala3.westeurope.cloudapp.azure.com/dashboard/projects/lampepfl/steps/builds/3.8.3-RC2%3A2026-03-10/logs)

Following code crashes the compiler 

## Compiler version

Last good release: 3.8.1-RC1-bin-20251221-327be51-NIGHTLY
First bad release: 3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY
Exact bisect failed, the first bad commit could be any of:
434941ab7ea47815a89c0e1b25e2d8ebcc54e68a 
910529a5d42880dd7c6ff63203b5d8fdafccb95c
Both are from https://github.com/scala/scala3/pull/24679

## Minimized code

```scala
import language.experimental.captureChecking

enum Result[+T, +E]:
  case Ok[+T](value: T) extends Result[T, Nothing]
  case Err[+E](error: E) extends Result[Nothing, E]

def catchException(body: => Int): Result[Int, Null] =
  try Result.Ok(body)
  catch case _: Throwable => Result.Err(null)
```

## Crash stacktrace

<details> 

```scala

  unhandled exception while running cc on /Users/wmazur/projects/scala/community-build3/test.scala

  An unhandled exception was thrown in the compiler.
  Please file a crash report here:
  https://github.com/scala/scala3/issues/new/choose
  For non-enriched exceptions, compile with -Xno-enrich-error-messages.


     while compiling: /Users/wmazur/projects/scala/community-build3/test.scala
        during phase: cc
                mode: Mode(ImplicitsEnabled)
     library version: version 3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY
    compiler version: version 3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY-git-082ba9e
            settings: -classpath /Users/wmazur/Library/Caches/Coursier/v1/https/repo.scala-lang.org/artifactory/maven-nightlies/org/scala-lang/scala3-library_3/3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY/scala3-library_3-3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY.jar:/Users/wmazur/Library/Caches/Coursier/v1/https/repo.scala-lang.org/artifactory/maven-nightlies/org/scala-lang/scala-library/3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY/scala-library-3.8.1-RC1-bin-20251222-082ba9e-NIGHTLY.jar -d /Users/wmazur/projects/scala/community-build3/.scala-build/community-build3_5ecdb296b2-9d23cdd7be/classes/main -sourceroot /Users/wmazur/projects/scala/community-build3

Exception in thread "main" java.lang.AssertionError: assertion failed
        at scala.runtime.Scala3RunTime$.assertFailed(Scala3RunTime.scala:13)
        at dotty.tools.dotc.cc.CaptureOps$package$.recur$1(CaptureOps.scala:232)
        at dotty.tools.dotc.cc.CaptureOps$package$.boxDeeply(CaptureOps.scala:255)
        at dotty.tools.dotc.cc.Setup.dotty$tools$dotc$cc$Setup$SetupTypeMap$$_$normalizeCaptures$$anonfun$1(Setup.scala:257)
        at scala.collection.immutable.List.loop$3(List.scala:466)
        at scala.collection.immutable.List.mapConserve(List.scala:491)
        at dotty.tools.dotc.cc.Setup$SetupTypeMap.normalizeCaptures(Setup.scala:257)
        at dotty.tools.dotc.cc.Setup$SetupTypeMap.normalizeCaptures$(Setup.scala:192)
        at dotty.tools.dotc.cc.Setup$toCapturing$2$.normalizeCaptures(Setup.scala:376)
        at dotty.tools.dotc.cc.Setup$toCapturing$2$.defaultApply(Setup.scala:434)
        at dotty.tools.dotc.cc.Setup$toCapturing$2$.innerApply(Setup.scala:470)
        at dotty.tools.dotc.cc.Setup$SetupTypeMap.apply(Setup.scala:206)
        at dotty.tools.dotc.cc.Setup$SetupTypeMap.apply$(Setup.scala:192)
        at dotty.tools.dotc.cc.Setup$toCapturing$2$.apply(Setup.scala:376)
        at dotty.tools.dotc.cc.Setup.transform$1(Setup.scala:474)
        at dotty.tools.dotc.cc.Setup.dotty$tools$dotc$cc$Setup$$transformExplicitType(Setup.scala:480)
        at dotty.tools.dotc.cc.Setup$$anon$2.transformTT(Setup.scala:520)
        at dotty.tools.dotc.cc.Setup$$anon$2.postProcess(Setup.scala:625)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:618)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
        at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1710)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:616)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:609)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:571)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1359)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1778)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:616)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.apply(Trees.scala:1819)
        at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1775)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:592)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1359)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1360)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverse(Trees.scala:1818)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.recur$2(tpd.scala:1357)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1362)
        at dotty.tools.dotc.ast.tpd$TreeTraverserWithPreciseImportContexts.apply(tpd.scala:1354)
        at dotty.tools.dotc.ast.Trees$Instance$TreeAccumulator.foldOver(Trees.scala:1784)
        at dotty.tools.dotc.ast.Trees$Instance$TreeTraverser.traverseChildren(Trees.scala:1820)
        at dotty.tools.dotc.cc.Setup$$anon$2.traverse(Setup.scala:616)
        at dotty.tools.dotc.cc.Setup.setupUnit(Setup.scala:940)
        at dotty.tools.dotc.cc.CheckCaptures$CaptureChecker.checkUnit(CheckCaptures.scala:2031)
        at dotty.tools.dotc.transform.Recheck.run(Recheck.scala:154)
        at dotty.tools.dotc.cc.CheckCaptures.run(CheckCaptures.scala:270)
        at dotty.tools.dotc.core.Phases$Phase.runOn$$anonfun$1(Phases.scala:406)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.immutable.List.foreach(List.scala:327)
        at dotty.tools.dotc.core.Phases$Phase.runOn(Phases.scala:399)
        at dotty.tools.dotc.transform.Recheck.runOn(Recheck.scala:158)
        at dotty.tools.dotc.Run.runPhases$1$$anonfun$1(Run.scala:380)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:15)
        at scala.runtime.function.JProcedure1.apply(JProcedure1.java:10)
        at scala.collection.ArrayOps$.foreach$extension(ArrayOps.scala:1327)
        at dotty.tools.dotc.Run.runPhases$1(Run.scala:373)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$2(Run.scala:420)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1$$anonfun$adapted$1(Run.scala:420)
        at scala.Function0.apply$mcV$sp(Function0.scala:45)
        at dotty.tools.dotc.Run.showProgress(Run.scala:482)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$1(Run.scala:420)
        at dotty.tools.dotc.Run.compileUnits$$anonfun$adapted$1(Run.scala:432)
        at dotty.tools.dotc.util.Stats$.maybeMonitored(Stats.scala:69)
        at dotty.tools.dotc.Run.compileUnits(Run.scala:432)
        at dotty.tools.dotc.Run.compileSources(Run.scala:319)
        at dotty.tools.dotc.Run.compile(Run.scala:304)
        at dotty.tools.dotc.Driver.doCompile(Driver.scala:38)
        at dotty.tools.dotc.Driver.process(Driver.scala:213)
        at dotty.tools.dotc.Driver.process(Driver.scala:181)
        at dotty.tools.dotc.Driver.process(Driver.scala:193)
        at dotty.tools.dotc.Driver.main(Driver.scala:223)
        at dotty.tools.dotc.Main.main(Main.scala)
```

</details>

## Expectation

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -1683,7 +1683,7 @@ trait Applications extends Compatibility {
       val widened = TypeComparer.dropTransparentTraits(
         tree.tpe.parents.reduceLeft(TypeComparer.andType(_, _)),
         pt)
-      if widened <:< pt then Typed(tree, TypeTree(widened))
+      if widened <:< pt then Typed(tree, TypeTree(widened, inferred = true))
       else tree
     else tree
 

--- a/tests/neg-custom-args/captures/effect-swaps-explicit.check
+++ b/tests/neg-custom-args/captures/effect-swaps-explicit.check
@@ -2,10 +2,15 @@
 63 |      Result:
 64 |        Future: // error, type mismatch
    |      ^
-   |      Found:    Result.Ok[Future[T^'s1]^{fr, contextual$1}]^'s2
-   |      Required: Result[Future[T], Nothing]
+   |Capability `contextual$1` outlives its scope: it leaks into outer capture set 's1 which is owned by method fail3.
+   |The leakage occurred when trying to match the following types:
    |
-   |      Note that capability `fr` cannot flow into capture set {}.
+   |Found:    (contextual$1: boundary.Label[Result[Future[T^'s2]^'s3, E^'s4]^'s5]^) ?->{Result, fr, async}
+   |  Result[Future[T^'s6]^{fr, contextual$1}, E^'s7]^'s8
+   |Required: (boundary.Label[Result[Future[T^'s9]^'s1, E^'s10]^'s11]^) ?=> Result[Future[T^'s9]^'s1, E^'s10]^'s11
+   |
+   |where:    ?=> refers to a root capability created in method fail3 when checking argument to parameter body of method apply
+   |          ^   refers to the root capability caps.any
    |
 65 |          fr.await.ok
    |--------------------------------------------------------------------------------------------------------------------
@@ -20,11 +25,11 @@
 -- [E007] Type Mismatch Error: tests/neg-custom-args/captures/effect-swaps-explicit.scala:69:10 ------------------------
 69 |          Future: fut ?=> // error, type mismatch
    |          ^
-   |Capability `contextual$9` outlives its scope: it leaks into outer capture set 's3 which is owned by method fail4.
+   |Capability `contextual$9` outlives its scope: it leaks into outer capture set 's12 which is owned by method fail4.
    |The leakage occurred when trying to match the following types:
    |
-   |Found:    (contextual$9: boundary.Label[Result[Future[T^'s4]^'s5, E^'s6]]^) ?->{fr, async} Future[T^'s7]^{fr, contextual$9}
-   |Required: (boundary.Label[Result[Future[T^'s8]^'s3, E^'s9]]^) ?=> Future[T^'s8]^'s3
+   |Found:    (contextual$9: boundary.Label[Result[Future[T^'s13]^'s14, E^'s15]]^) ?->{fr, async} Future[T^'s16]^{fr, contextual$9}
+   |Required: (boundary.Label[Result[Future[T^'s17]^'s12, E^'s18]]^) ?=> Future[T^'s17]^'s12
    |
    |where:    ?=> refers to a root capability created in method fail4 when checking argument to parameter body of method make
    |          ^   refers to the root capability caps.any
@@ -35,7 +40,7 @@
 -- [E007] Type Mismatch Error: tests/neg-custom-args/captures/effect-swaps-explicit.scala:74:10 ------------------------
 74 |          Future: fut ?=>  // error: type mismatch
    |          ^
-   |          Found:    Future[T^'s10]^{fr, lbl}
+   |          Found:    Future[T^'s19]^{fr, lbl}
    |          Required: Future[T]
    |
    |          Note that capability `fr` cannot flow into capture set {}.

--- a/tests/neg-custom-args/captures/effect-swaps.check
+++ b/tests/neg-custom-args/captures/effect-swaps.check
@@ -2,10 +2,15 @@
 63 |      Result:
 64 |        Future: // error, type mismatch
    |      ^
-   |      Found:    Result.Ok[Future[T^'s1]^{fr, contextual$1}]^'s2
-   |      Required: Result[Future[T], Nothing]
+   |Capability `contextual$1` outlives its scope: it leaks into outer capture set 's1 which is owned by method fail3.
+   |The leakage occurred when trying to match the following types:
    |
-   |      Note that capability `fr` cannot flow into capture set {}.
+   |Found:    (contextual$1: boundary.Label[Result[Future[T^'s2]^'s3, E^'s4]^'s5]^) ?->{Result, fr, async}
+   |  Result[Future[T^'s6]^{fr, contextual$1}, E^'s7]^'s8
+   |Required: (boundary.Label[Result[Future[T^'s9]^'s1, E^'s10]^'s11]^) ?=> Result[Future[T^'s9]^'s1, E^'s10]^'s11
+   |
+   |where:    ?=> refers to a root capability created in method fail3 when checking argument to parameter body of method apply
+   |          ^   refers to the root capability caps.any
    |
 65 |          fr.await.ok
    |--------------------------------------------------------------------------------------------------------------------
@@ -20,11 +25,11 @@
 -- [E007] Type Mismatch Error: tests/neg-custom-args/captures/effect-swaps.scala:69:10 ---------------------------------
 69 |          Future: fut ?=> // error, type mismatch
    |          ^
-   |Capability `contextual$9` outlives its scope: it leaks into outer capture set 's3 which is owned by method fail4.
+   |Capability `contextual$9` outlives its scope: it leaks into outer capture set 's12 which is owned by method fail4.
    |The leakage occurred when trying to match the following types:
    |
-   |Found:    (contextual$9: boundary.Label[Result[Future[T^'s4]^'s5, E^'s6]]^) ?->{fr, async} Future[T^'s7]^{fr, contextual$9}
-   |Required: (boundary.Label[Result[Future[T^'s8]^'s3, E^'s9]]^) ?=> Future[T^'s8]^'s3
+   |Found:    (contextual$9: boundary.Label[Result[Future[T^'s13]^'s14, E^'s15]]^) ?->{fr, async} Future[T^'s16]^{fr, contextual$9}
+   |Required: (boundary.Label[Result[Future[T^'s17]^'s12, E^'s18]]^) ?=> Future[T^'s17]^'s12
    |
    |where:    ?=> refers to a root capability created in method fail4 when checking argument to parameter body of method make
    |          ^   refers to the root capability caps.any
@@ -35,7 +40,7 @@
 -- [E007] Type Mismatch Error: tests/neg-custom-args/captures/effect-swaps.scala:74:10 ---------------------------------
 74 |          Future: fut ?=>  // error: type mismatch
    |          ^
-   |          Found:    Future[T^'s10]^{fr, lbl}
+   |          Found:    Future[T^'s19]^{fr, lbl}
    |          Required: Future[T]
    |
    |          Note that capability `fr` cannot flow into capture set {}.

--- a/tests/pos-custom-args/captures/i25497.scala
+++ b/tests/pos-custom-args/captures/i25497.scala
@@ -1,0 +1,7 @@
+enum Result[+T, +E]:
+  case Ok[+T](value: T) extends Result[T, Nothing]
+  case Err[+E](error: E) extends Result[Nothing, E]
+
+def catchException(body: => Int): Result[Int, Null] =
+  try Result.Ok(body)
+  catch case _: Throwable => Result.Err(null)

